File: packages/oz.ozhtml/src/lexer.js
const tokenConst = require("./token_const");
const charConst = require("./char_const");
const {HtmlToken} = require("./token");
const HtmlLexer = function (input){
    let CHAR=-1;
    let me = this;
    let isWord=function(char){
        return ( 96 < char && 123 > char) || ( 64 < char && 91 > char || char === charConst.MINUS || char === 95 || char === 35);
    };
    let isNumeric=function(char){
        return ( 47 < char && 58 > char);
    };
    let isWhiteSpace=function(char){
        return char === 32 || char === 13 || char === 10;
    };
    me.tokens=[];
    me.tokenize =function () {
        if (me.tokens.length)
            return me.tokens.pop();
        let char;
        if (CHAR!==-1) {
            char = CHAR;
            CHAR=-1;
        }else {
            char = input.read();
        }
        if (char===charConst.GREATER){
            char = input.read();
            if (char===charConst.BACKSLASH){
                return new HtmlToken(tokenConst.TAGHEADCLOSE)
            }else if (char===charConst.BANG){
                 input.read();
                 input.read();
                let ii=0;
                let value="";
                while(true){
                    char= input.read();
                    value+= String.fromCharCode(char);
                    if (char===charConst.EOF){
                        break;
                    }else if (ii===0&&char===charConst.MINUS){
                        ii=1;
                    }else if (ii===1&&char===charConst.MINUS){
                        ii=2;
                    }else if (ii===2&&char===charConst.MINOR){
                        break;
                    }else {
                        ii=0;
                    }
                }
                value =value.substring(0,value.length-3);
                return new HtmlToken(tokenConst.TAGCOMMENT,value);
            }else {
                CHAR = char;
                return new HtmlToken(tokenConst.TAGHEADOPEN);
            }
        }else  if (char===charConst.BACKSLASH){
            char = input.read();
            if (char===charConst.MINOR){
                console.log(char);
            }
            return new HtmlToken(tokenConst.TAGHEADCLOSE);
        }else if (char===charConst.MINOR){
            let value="";
            char = input.read();
            while (char !== charConst.GREATER && char !== charConst.EOF){
                value+= String.fromCharCode(char);
                char = input.read();
            }
            CHAR = char;
            return new HtmlToken(tokenConst.TAGBODY,value);
        }else if (char===charConst.PARENTHESISOPEN){
            return new HtmlToken(tokenConst.PARENTHESISOPEN);
        }else if (char===charConst.PARENTHESISCLOSE){
            return new HtmlToken(tokenConst.PARENTHESISCLOSE);
        }else if (char===charConst.EQUAL) {
            return new HtmlToken(tokenConst.EQUAL);
        }else if (char===charConst.MULTIPLICATION) {


            return new HtmlToken(tokenConst.MULTIPLICATION);

        }else if (char===charConst.BRACKETOPEN) {
            return new HtmlToken(tokenConst.BRACKETOPEN);
        }else if (char===charConst.BRACKETClOSE) {
            return new HtmlToken(tokenConst.BRACKETCLOSE);
        }else if (char===charConst.DOT) {
            return new HtmlToken(tokenConst.DOT);
        }else if (char===charConst.DOUBLEDOT) {
            return new HtmlToken(tokenConst.DOUBLEDOT);
        }else if (char===charConst.DOUBLEQUOTES) {
            let value="";
            char = input.read();
            while (char !== charConst.DOUBLEQUOTES && char !== charConst.EOF){
                value+= String.fromCharCode(char);
                char = input.read();
            }
            return new HtmlToken(tokenConst.STRING,value);
        }else if (isWhiteSpace(char)) {
            while (isWhiteSpace(char)){
                char = input.read();
            }
            CHAR = char;
            return new HtmlToken(tokenConst.WHITESPACE);
        }else if (isWord(char)) {
            let value="";
            while (isWord(char)||isNumeric(char)){
                value+= String.fromCharCode(char);
                char = input.read();
            }
            CHAR = char;
            return new HtmlToken(tokenConst.WORD,value);
        }else if (isNumeric(char)) {
            let value="";
            while (isNumeric(char)){
                value+= String.fromCharCode(char);
                char = input.read();
            }
            CHAR = char;
            return new HtmlToken(tokenConst.NUMERIC,value);
        }else if (char===charConst.EOF) {
            return new HtmlToken(tokenConst.EOF);
        }else {
            throw "do'nt know charcode "+char;
        }

    };
};
module.exports = {
    HtmlLexer
};

File: go/src/github.com/w3esoft/oz/ozhtml/lexer/lexer.go
package lexer

import (
	"github.com/w3esoft/glaive/input"
	"github.com/w3esoft/oz/ozhtml/token"
	"strings"
)

const (
	DOUBLEDOT        = 58
	GREATER          = 60
	MINOR            = 62
	EQUAL            = 61
	DOUBLEQUOTES     = 34
	MULTIPLICATION   = 42
	BRACKETOPEN      = 91
	BRACKETClOSE     = 93
	PARENTHESISOPEN  = 40
	PARENTHESISCLOSE = 41
	BACKSLASH        = 47
	DOT              = 46
	BANG             = 33
	MINUS            = 45
	UNDERSCORE       = 95
	HASH             = 35
	EOF              = 0
)
const (
	MODE_TAG  = 1
	MODE_BODY = 2
)

type Lexer struct {
	CHAR   int
	Input  input.Input
	Tokens [] *token.Token
	pos    int
	mode   int
}

func New(input input.Input) *Lexer {
	lexer := Lexer{}
	lexer.CHAR = -1
	lexer.pos = 0
	lexer.mode = MODE_BODY;
	lexer.Input = input
	return &lexer
}

func (lexer *Lexer) Read() uint8 {
	var char uint8
	if lexer.CHAR != -1 {
		char = uint8(lexer.CHAR)
		lexer.CHAR = -1
	} else {
		lexer.pos++
		char = lexer.Input.Read()

	}
	return char
}
func (lexer *Lexer) Pos() int {
	if lexer.CHAR == -1 {
		return lexer.pos
	} else {
		return lexer.pos - 1
	}
}
func (lexer *Lexer) Tokenize() *token.Token {
	if len(lexer.Tokens) > 0 {
		r := lexer.Tokens[len(lexer.Tokens)-1:][0]
		lexer.Tokens = lexer.Tokens[:len(lexer.Tokens)-1]
		return r
	}
	offset := lexer.Pos()
	var value *string = nil
	char := lexer.Read()
	if char == EOF {
		position := &token.Position{Len: 0, Offset: offset}
		v := string(char)
		value = &v
		return token.New(token.EOF, value, position, true)
	} else if lexer.mode == MODE_BODY {
		switch {
		case char == GREATER:
			char = lexer.Read()
			switch char {
			case BANG:
				char = lexer.Read()
				if (IsWord(char)) && char != MINUS {
					var v = ""
					for IsWord(char) {
						v = v + string(char)
						char = lexer.Read()
					}
					if strings.Compare(strings.ToLower(v), "doctype") != 0 {
						len := lexer.Pos() - offset
						position := &token.Position{Len: len, Offset: offset}
						return token.New(token.DOCTYPE_OPEN, value, position, false)
					}
					lexer.CHAR = int(char)
					len := lexer.Pos() - offset
					position := &token.Position{Len: len, Offset: offset}
					lexer.mode = MODE_TAG;
					return token.New(token.DOCTYPE_OPEN, value, position, true)
				}
				if char != MINUS {
					len := lexer.Pos() - offset
					position := &token.Position{Len: len, Offset: offset}
					return token.New(token.COMMENT, value, position, false)
				}
				char = lexer.Read()
				if char != MINUS {
					len := lexer.Pos() - offset
					position := &token.Position{Len: len, Offset: offset}
					return token.New(token.COMMENT, value, position, false)
				}

				var ii = 0
				var v = ""
				for {
					v = v + string(char)
					char = lexer.Read()
					b := false;
					switch {
					case char == EOF:
						len := lexer.Pos() - offset
						position := &token.Position{Len: len, Offset: offset}
						return token.New(token.COMMENT, value, position, false)
					case ii == 0 && char == MINUS:
						ii = 1
					case ii == 1 && char == MINUS:
						ii = 2
					case ii == 2 && char == MINOR:
						b = true;
					default:
						ii = 0
					}
					if (b) {
						break
					}

				}
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.COMMENT, value, position, true)
			case BACKSLASH:
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				lexer.mode = MODE_TAG;
				return token.New(token.TAG_END_OPEN, value, position, true)
			default:
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				lexer.mode = MODE_TAG;
				return token.New(token.TAG_OPEN, value, position, true)

			}
		case char == EOF:
			position := &token.Position{Len: 0, Offset: offset}
			return token.New(token.EOF, value, position, true)
		default:
			var v = string(char)
			char = lexer.Read();
			for char != GREATER && char != EOF {
				v = v + string(char)
				char = lexer.Read()
			}
			lexer.CHAR = int(char)
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TEXT, value, position, true)
		}
	} else if lexer.mode == MODE_TAG {
		switch {
		case char == MINOR:
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			lexer.mode = MODE_BODY
			return token.New(token.TAG_CLOSE, value, position, true)
		case char == BACKSLASH:
			char = lexer.Read()
			if char == MINOR {
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				lexer.mode = MODE_BODY
				return token.New(token.TAG_END_CLOSE, value, position, true)
			} else {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				lexer.mode = MODE_BODY
				return token.New(token.TAG_END_CLOSE, value, position, false)
			}
		case IsWord(char) || char == UNDERSCORE || HASH == char:
			v := ""
			if !(IsWord(char) || char == UNDERSCORE || HASH == char) {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.KEY, value, position, false)
			}
			v = v + string(char)
			char = lexer.Read()
			for IsWord(char) || char == MINUS || char == UNDERSCORE || char == DOT || char == DOUBLEDOT || IsNumeric(char) {
				v = v + string(char)
				char = lexer.Read()
			}
			lexer.CHAR = int(char)
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.KEY, value, position, true)
		case char == MULTIPLICATION:
			char = lexer.Read()
			v := ""
			if !(IsWord(char) || char == UNDERSCORE || HASH == char ) {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.TAG_KEY_OUTPUT, value, position, false)
			}
			v = v + string(char)
			char = lexer.Read()
			for IsWord(char) || char == MINUS || char == UNDERSCORE || char == DOT || char == DOUBLEDOT || IsNumeric(char) {
				v = v + string(char)
				char = lexer.Read()
			}
			lexer.CHAR = int(char)
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TAG_KEY_OUTPUT, value, position, true)
		case char == BRACKETOPEN:
			v := ""
			char = lexer.Read()
			if !(IsWord(char) || char == UNDERSCORE || HASH == char) {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.TAG_KEY_INPUT, value, position, false)
			}
			v = v + string(char)
			char = lexer.Read()
			for IsWord(char) || char == MINUS || char == UNDERSCORE || char == DOT || char == DOUBLEDOT || IsNumeric(char) || IsWhiteSpace(char) {
				v = v + string(char)
				char = lexer.Read()
			}
			if char != BRACKETClOSE {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.TAG_KEY_INPUT, value, position, false)
			}
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TAG_KEY_INPUT, value, position, true)
		case char == PARENTHESISOPEN:
			v := ""
			char = lexer.Read()
			if !(IsWord(char) || char == UNDERSCORE || HASH == char) {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.TAG_KEY_OUTPUT, value, position, false)
			}
			v = v + string(char)
			char = lexer.Read()
			for IsWord(char) || char == MINUS || char == UNDERSCORE || char == DOT || char == DOUBLEDOT || IsNumeric(char) || IsWhiteSpace(char) {
				v = v + string(char)
				char = lexer.Read()
			}
			if char != PARENTHESISCLOSE {
				lexer.CHAR = int(char)
				len := lexer.Pos() - offset
				position := &token.Position{Len: len, Offset: offset}
				return token.New(token.TAG_KEY_OUTPUT, value, position, false)
			}
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TAG_KEY_OUTPUT, value, position, true)
		case char == EQUAL:
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			return token.New(token.TAG_EQUAL, value, position, true)
		case char == DOUBLEQUOTES:
			v := ""
			char = lexer.Read()
			for char != DOUBLEQUOTES {
				if char == EOF {
					lexer.CHAR = int(char)
					len := lexer.Pos() - offset
					position := &token.Position{Len: len, Offset: offset}
					value = &v
					return token.New(token.TAG_VALUE_STRING, value, position, false)
				}
				v = v + string(char)
				char = lexer.Read()

			}
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TAG_VALUE_STRING, value, position, true)
		case IsWhiteSpace(char):
			for IsWhiteSpace(char) {
				char = lexer.Read()
			}
			lexer.CHAR = int(char)
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			return token.New(token.TAG_WHITESPACE, value, position, true)
		case IsNumeric(char):
			v := ""
			for IsNumeric(char) {
				v = v + string(char)
				char = lexer.Read()
			}
			lexer.CHAR = int(char)
			len := lexer.Pos() - offset
			position := &token.Position{Len: len, Offset: offset}
			value = &v
			return token.New(token.TAG_VALUE_NUMERIC, value, position, true)

		default:
		}
	}
	len := lexer.Pos() - offset
	position := &token.Position{Len: len, Offset: offset}
	v := string(char)
	value = &v
	return token.New(token.INVALID, value, position, false)
}
func (l *Lexer) TokenPush(i2 *token.Token) {
	l.Tokens = append(l.Tokens, i2)
}

func IsWord(char uint8) bool {
	return (96 < char && 123 > char) || (64 < char && 91 > char)
}
func IsNumeric(char uint8) bool {
	return 47 < char && 58 > char
}

func IsWhiteSpace(char uint8) bool {
	return char == 32 || char == 13 || char == 10 || char == 9
}
